### DISCUSSION TITLE
Node maintenance expectations

### DISCUSSION THREAD

**Discussion by craigcabrey** [rook/rook#14670] [Q&A]
Hi!

Thanks for building rook. It hasn't been a completely smooth journey but I'm getting there. I haven't found any clear documentation (except for one off comments in some spots) around node maintenance.

Specifically, I'm trying to understand what the expectations are around node drains. When a node is drained (via `kubectl drain` or an API call), does the operator act on that? If so, is there any documentation around expected behavior?

Thanks!

**Comment by travisn** [comment]
The node drain is described best in the [PDB design doc](https://github.com/rook/rook/blob/master/design/ceph/ceph-managed-disruptionbudgets.md) for OSDs. Other daemons don't have special drain behavior. There is just a brief mention of this behavior in the cephCluster settings for [disruptionManagement](https://rook.io/docs/rook/latest/CRDs/Cluster/ceph-cluster-crd/#cluster-settings):
_______________

`managePodBudgets`: if true, the operator will create and manage PodDisruptionBudgets for OSD, Mon, RGW, and MDS daemons. OSD PDBs are managed dynamically via the strategy outlined in the [design](https://github.com/rook/rook/blob/master/design/ceph/ceph-managed-disruptionbudgets.md). The operator will block eviction of OSDs by default and unblock them safely when drains are detected.

**Reply by craigcabrey** [reply]
Thanks. I do have `managePodBudgets` enabled, but the operator does not seem to detect drains:

```
2024-09-07 18:53:10.445720 D | clusterdisruption-controller: deleting default pdb with maxUnavailable=1 for all osd
2024-09-07 18:53:13.314341 D | clusterdisruption-controller: reconciling "rook-ceph/"
2024-09-07 18:53:13.902933 D | clusterdisruption-controller: pg health check timeout is not set in the cluster. waiting for PGs to get active+clean
2024-09-07 18:53:13.902970 I | clusterdisruption-controller: all "host" failure domains: [ms01-node-1 ms01-node-2 ms01-node-3 ms01-node-4 ms01-node-5 ms01-node-6 ms01-node-7]. osd is down in failure domain: "". active node drains: false. pg health: "cluster is not fully clean. PGs: [{StateName:active+clean Count:407} {StateName:active+remapped+backfill_wait Count:11} {StateName:active+remapped+backfilling Count:10} {StateName:active+clean+scrubbing+deep Count:4} {StateName:active+clean+scrubbing Count:1}]"
2024-09-07 18:53:14.469005 D | clusterdisruption-controller: deleting default pdb with maxUnavailable=1 for all osd
```

It does complain about an unclean cluster (it is rebalancing), so maybe that has something to do with it. However, the `active node drains: false` sticks out to me as false.

Is there anyway to debug why the operator isn't detecting an ongoing drain?

**Reply by travisn** [reply]
@sp98 thoughts?

**Reply by sp98** [reply]
`active node drains: false`
This could happen in the scenario when the drained node is added back (uncordoned) and the OSD has restarted on it but the PGs are still not active+clean (rebalancing)


Expectation of node drain behavior  in rook with respect to OSDs. 

- Consider a cluster with 3 OSDs on three Nodes. Failure domain is `Node`. 
- By default, Rook will allow 1 OSD to be drained in a failure domain by creating a `default PDB` with `maxUnavailable = 1`
- User drains Node-1.  Rook will allow it. But rook will delete the `default PDB` and create two new PDBs on each of the other nodes (Node-2 and Node-3). These will be `blocking PDBs` because `maxUnavailable` will be set to 0 for these PDBs.  (The idea here is to prevent OSDs to go down in multiple failures domains at the same time). 
- Node-1 maintenance is complete and user uncordons Node-1.  Drained OSD will be added back to this node. Rook will wait till the data is rebalanced ( that's PGs are active+clean) before deleting the `blocking PDBs` from Node-2 and Node-3 and adding the `default PDB` again. 

This is explained in more detail [here](https://github.com/rook/rook/blob/master/design/ceph/ceph-managed-disruptionbudgets.md).

**Reply by craigcabrey** [reply]
Thanks for the response. I ended up upgrading to the new operator (1.15.1), and did another test drain. The operator picked up that drain just fine, so maybe just a bug.
